src/pybind/mgr/volumes/fs/nfs.py

```diff
@@ -0,0 +1,287 @@
+            if self.rados_namespace:
+                ioctx.set_namespace(self.rados_namespace)
+            ioctx.remove_object("export-{}".format(export_id))
+
+    def _save_export(self, export):
+        export.fsal.cephx_key = self.cephx_key
+        self.exports[export.export_id] = export
+        conf_block = export.to_export_block(self.export_defaults)
+        self._write_raw_config(conf_block, "export-{}".format(export.export_id))
+        self._persist_daemon_configuration()
+
+    def create_export(self, ex_dict):
+        ex_id = self._gen_export_id()
+        export = Export.from_dict(ex_id, ex_dict)
+        self._save_export(export)
+        return ex_id
+
+    def remove_export(self, export_id):
+        if export_id not in self.exports:
+            return None
+        export = self.exports[export_id]
+        del self.exports[export_id]
+        self._delete_export(export_id)
+        return export
+
+    def has_export(self, export_id):
+        return export_id in self.exports
+
+    def list_daemons(self):
+        return [daemon_id for daemon_id in self.daemons_conf_blocks]
+
+    def reload_daemons(self, daemons):
+        with self.mgr.rados.open_ioctx(self.rados_pool) as ioctx:
+            if self.rados_namespace:
+                ioctx.set_namespace(self.rados_namespace)
+            for daemon_id in daemons:
+                ioctx.notify("conf-{}".format(daemon_id))
+
+class NFSConfig(object):
+    exp_num = 0
+
+    def __init__(self, mgr, cluster_id):
+        self.cluster_id = "ganesha-%s" % cluster_id
+        self.pool_name = 'nfs-ganesha'
+        self.pool_ns = cluster_id
+        self.mgr = mgr
+        self.ganeshaconf = ''
+        self.key = ''
+
+    def update_user_caps(self):
+        if NFSConfig.exp_num > 0:
+            ret, out, err = self.mgr.mon_command({
+                'prefix': 'auth caps',
+                'entity': "client.%s" % (self.cluster_id),
+                'caps' : ['mon', 'allow *', 'osd', 'allow * pool=%s namespace=%s, allow rw tag cephfs data=a' % (self.pool_name, self.pool_ns), 'mds', 'allow * path=/'],
+                })
+
+            if ret!= 0:
+                return ret, out, err
+
+    def create_common_config(self, nodeid):
+        result = "NFS_CORE_PARAM {\n Enable_NLM = false;\n Enable_RQUOTA = false;\n Protocols = 4;\n}\n\n"
+        result += "CACHEINODE {\n Dir_Chunk = 0;\n NParts = 1;\n Cache_Size = 1;\n}\n\n"
+        result += "NFSv4 {\n RecoveryBackend = rados_cluster;\n Minor_Versions = 1, 2;\n}\n\n"
+        result += "RADOS_URLS {{\n userid = {};\n}}\n\n".format(self.cluster_id)
+        #result += "%url rados://{}/{}/{}\n\n".format(self.pool_name, self.pool_ns, nodeid)
+        result += "%url rados://{}/{}/export-1\n\n".format(self.pool_name, self.pool_ns)
+        result += "RADOS_KV {{\n pool = {};\n namespace = {};\n UserId = {};\n nodeid = {};\n}}\n\n".format(self.pool_name, self.pool_ns, self.cluster_id, nodeid)
```

Comment:
This is really hard to read and will be hard to change in the future. It should probably be laid out in a more readable fashion (maybe using triple quotes?).